### PR TITLE
crossdock/client: refactor

### PR DESCRIPTION
**Pull request by abhinav**
I sort of went overboard implementing this originally and ended up with something really hard to follow.

This simplifies things: We still have something that can record multiple success/failure/skip outputs, allowing writing behaviors similar to how we write tests, but it's no longer hidden three layers deep.

Behaviors are executed inside a goroutine so that we can do `runtime.Goexit()` for irrecoverable failures. This is similar to the technique used by the `testing` package.

### CODE REVIEW

**Comment by abhinav** [comment]
CC @yarpc/golang

**Comment by breerly** [comment]
I'd like to see the 3 echo behaviors broken out into their own functions named after the crossdock behavior if possible, but other than that this is a big improvement :+1:

**Comment by abhinav** [comment]
Can you clarify on what you mean? They are already three separate functions named EchoRaw, EchoJSON and EchoThrift.

**Comment by breerly** [comment]
Look how Python organizes them - https://github.com/yarpc/yarpc-python/tree/master/crossdock/async

**Comment by breerly** [comment]
It's much easier to seek to a particular behavior that way - and will stay that way as we end up with dozens of behaviors

**Comment by abhinav** [comment]
Discussed offline. Changes incoming.

File: crossdock/client/behavior/run.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+//
+// Functions like Fatalf won't work if the behavior is not executed inside a
+// Run context.
+func Run(f func()) {
+	done := make(chan struct{})
+
+	// We run the function inside a goroutine so that Fatalf can simply call
+	// runtime.Goexit to stop execution.
+	go func() {
+		defer func() {
```
Comment (prashantv) [inline crossdock/client/behavior/run.go]: can avoid inline function for the channel write by using `close` instead. i actually prefer close for cases like this:
 - releases resources for the channel immediately
 - panics if there's a code bug which causes it to try to close twice

```go
go func() {
  defer close(done)
  f()
}()
```

File: crossdock/client/random/rand.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+package random
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"io"
+	"io/ioutil"
+)
+
+// Bytes generates the gvien number of random bytes.
```
Comment (prashantv) [inline crossdock/client/random/rand.go]: typo, `s/gvien/given`

File: crossdock/client/random/rand.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	return out
+}
+
+// String generates a random string of the given size.
+func String(length int64) string {
+	bs, err := ioutil.ReadAll(io.LimitReader(rand.Reader, length))
+	if err != nil {
+		panic(err)
+	}
+	return base64.RawStdEncoding.EncodeToString(bs)
```
Comment (prashantv) [inline crossdock/client/random/rand.go]: `base64` takes up more space than the bytes, so if you get n bytes and encode it, the string length > n.

you can probably use `base64.DecodedLen` to figure out how many bytes you really need. (optionally, write a simple unit test to verify that a few calls to this really give you a string of expected length)

File: crossdock/client/random/rand.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	out := make([]byte, length)
+	if _, err := io.ReadFull(rand.Reader, out); err != nil {
+		panic(err)
+	}
+	return out
+}
+
+// String generates a random string of the given size.
+func String(length int64) string {
+	bs, err := ioutil.ReadAll(io.LimitReader(rand.Reader, length))
```
Comment (prashantv) [inline crossdock/client/random/rand.go]: just use `Bytes` to get the random bytes

File: crossdock/client/echo/behavior.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+		Server:    s.Server,
+	})
+}
+
+// createEchoSink wraps a Sink to have transport, encoding, and server
+// information.
+func createEchoSink(encoding string, s behavior.Sink, p behavior.Params) behavior.Sink {
+	return echoSink{
+		Sink:      s,
+		Transport: p.Param("transport"),
```
Comment (prashantv) [inline crossdock/client/echo/behavior.go]: will there be a known list of params? if so, maybe we should use a custom type and have all the known params defined somewhere?

otherwise, we should create constants for the string keys

File: crossdock/client/client.go [hunk 6]
```diff
@@ -36,19 +38,24 @@ func behaviorRequestHandler(w http.ResponseWriter, r *http.Request) {
...
-	// All behaviors will eventually contribute to this behavior's output.
-	bt := BehaviorTester{Params: httpParams{Request: r}}
-	v := bt.Param("behavior")
-	switch v {
-	case "raw", "json", "thrift":
-		runEchoBehavior(&bt, v)
-	default:
-		bt.NewBehavior(BasicEntryBuilder).Skipf("unknown behavior %v", v)
+	var s behavior.EntrySink
+	behavior.Run(func() { dispatch(&s, httpParams{r}) })
```
Comment (prashantv) [inline crossdock/client/client.go]: one thing that would be nice -- if the function panics for some reason, instead of crashing the process, we should recover, and log the panic as a fatalf (this is what the test framework does with test functions).

File: crossdock/client/behavior/sink.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+//////////////////////////////////////////////////////////////////////////////
+
+// EntrySink is a sink that keeps track of entries in-order
+type EntrySink struct{ entries []interface{} }
+
+// Put an entry into the EntrySink.
+func (e *EntrySink) Put(v interface{}) {
+	e.entries = append(e.entries, v)
```
Comment (prashantv) [inline crossdock/client/behavior/sink.go]: why is `v` an interface{}, is it not always `Entry`?

Comment (abhinav) [inline crossdock/client/behavior/sink.go]: So that we can emit more complex shapes in the output besides just `Entry`

File: crossdock/client/behavior/run.go (hunk 1 above)
Comment (abhinav) [inline crossdock/client/behavior/run.go]: :+1:

File: crossdock/client/random/rand.go (hunk 2 above)
Comment (abhinav) [inline crossdock/client/random/rand.go]: :+1:

File: crossdock/client/random/rand.go (hunk 3 above)
Comment (abhinav) [inline crossdock/client/random/rand.go]: Better yet: I can just slice the resulting string down to the actual requested size.

File: crossdock/client/random/rand.go (hunk 4 above)
Comment (abhinav) [inline crossdock/client/random/rand.go]: True. IDK why this was being duplicated.

File: crossdock/client/echo/behavior.go (hunk 5 above)
Comment (abhinav) [inline crossdock/client/echo/behavior.go]: Each behavior has its own configuration params. Creating constants for this makes sense.

File: crossdock/client/client.go (hunk 6 above)
Comment (abhinav) [inline crossdock/client/client.go]: :+1: 

To make that happen, I had to change `Run` to have information about the sink so the interface is now:

    Run(func(Sink)) []interface{} { ... }

So Run construct the sink and returns the entries logged by the function, and it handles panics.

**Comment by abhinav** [comment]
Updated

File: crossdock/client/echo/json.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+	ctx, _ := context.WithTimeout(context.Background(), time.Second)
+
+	var response jsonEcho
+	token := random.String(5)
+	_, err := client.Call(
+		&json.Request{Context: ctx, Procedure: "echo"},
+		&jsonEcho{Token: token},
+		&response,
+	)
+
```
Comment (prashantv) [inline crossdock/client/echo/json.go]: nit: delete blank

File: crossdock/client/echo/thrift.go [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+	client := echoclient.New(rpc.Channel("yarpc-test"))
+	ctx, _ := context.WithTimeout(context.Background(), time.Second)
+
+	token := random.String(5)
+	pong, _, err := client.Echo(
+		&thrift.Request{Context: ctx},
+		&echo.Ping{Beep: token},
+	)
+
```
Comment (prashantv) [inline crossdock/client/echo/thrift.go]: nit: delete blank line

**Comment by prashantv** [comment]
lgtm

File: crossdock/client/echo/thrift.go (hunk 9 above)
Comment (abhinav) [inline crossdock/client/echo/thrift.go]: :+1:

File: crossdock/client/echo/json.go (hunk 8 above)
Comment (abhinav) [inline crossdock/client/echo/json.go]: :+1:
